lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -592,27 +655,39 @@ monitorStakePools tr gp nl DBLayer{..} =
 monitorMetadata
     :: Tracer IO StakePoolLog
     -> GenesisParameters
-    -> (   PoolId
-        -> StakePoolMetadataUrl
-        -> StakePoolMetadataHash
-        -> IO (Maybe StakePoolMetadata)
-       )
     -> DBLayer IO
+    -> MVar Bool
     -> IO ()
-monitorMetadata tr gp fetchMetadata DBLayer{..} = forever $ do
-    refs <- atomically (unfetchedPoolMetadataRefs 100)
-    successes <- fmap catMaybes $ forM refs $ \(pid, url, hash) -> do
-        fetchMetadata pid url hash >>= \case
-            Nothing -> Nothing <$ do
-                atomically $ putFetchAttempt (url, hash)
-
-            Just meta -> Just hash <$ do
-                atomically $ putPoolMetadata hash meta
-
-    when (null refs || null successes) $ do
-        traceWith tr $ MsgFetchTakeBreak blockFrequency
-        threadDelay blockFrequency
+monitorMetadata tr gp DBLayer{..} stop = do
+    settings <- atomically readSettings
+    loop settings
   where
+    loop settings = do
+        (refs, successes) <- case poolMetadataSource settings of
+            FetchNone -> pure ([], [])
+            FetchDirect -> (fetcher [identityUrlBuilder]) >>= fetchThem
+            FetchSMASH (unSmashServer -> uri) -> (fetcher [registryUrlBuilder uri]) >>= fetchThem
+
+        when (null refs || null successes) $ do
+            traceWith tr $ MsgFetchTakeBreak blockFrequency
+            threadDelay blockFrequency
+        readMVar stop >>= \case
+            True -> pure ()
+            False -> loop settings
```

Comment:
It doesn't seem right that this re-implement the logic for blocking on a variable since we get this for free by using mvar. 
i.e.

by putting the _settings_ themselves in the mvar, and making the `loop` take settings only from that mvar, we ensure synchronization around the data that needs to be synchronize instead of creating additional logic to synchronize the access to the data.

Reply:
> It doesn't seem right that this re-implement the logic for blocking on a variable since we get this for free by using mvar.

I cannot follow. We're not blocking here at all. The stop mvar is *never* empty.

> by putting the settings themselves in the mvar

I don't like putting settings into an MVar. This data comes from the database already.

Reply:
> I don't like putting settings into an MVar. This data comes from the database already.

But that's the data for which we ensure to ensure the synchronization! So instead of dynamically controlling the behavior of the `loop` to make it do nothing or fetch, it is much easier (and far less confusing) to use an MVar to synchronize the settings and make the loop block when the settings are being modified. 

"I don't like it" is a very weak argument :grimacing: 

If you want to make sure *something* is only used by one thread at a time, put that one thing in an MVar. Don't build extra logic around it, the MVar does it well all alone already.

Reply:
> But that's the data for which we want to ensure the synchronization! So instead of dynamically controlling the behavior of the loop to make it do nothing or fetch

That's not what happens. The thread *terminates*... it doesn't do nothing. So you're proposing something completely different. And I haven't seen an argument what's wrong with my approach. This is basically how graceful thread shutdown works.

Reply:
>  haven't seen an argument what's wrong with my approach.

The argument is the complexity overhead that comes with managing the thread life-cycle. Since there's no need to start and restart a new thread, having to use two MVar to control the execution is complex and could be easily avoided. To be 100% clear, I am not saying that this current solution doesn't work, I am saying that there's a simpler solution which doesn't induce as much complexity (which is good for maintainers). What the loop does for fetching data is all clear now, it won't be in a month. So the simpler, the better. 

Don't get too attached to code, we throw code away every day. That's what the "soft" part in software is for :upside_down_face:

Reply:
I believe thread restarting is the superior approach. And I forgot to mention that it's the only approach that cleanly allows "read after write" and returns a true 200. With the other approaches, you either:

* risk a race condition if you drop the table in `pushSettings`
* can only return 202, because you have no knowledge when the setting was actually applied

I think 200 is far better for daedalus to consume as an API user.